README.md
# poeditor-service-provider

[![npm](https://img.shields.io/npm/v/poeditor-service-provider.svg)](https://www.npmjs.com/package/poeditor-service-provider)
[![CircleCI](https://circleci.com/gh/kazupon/poeditor-service-provider.svg?style=svg)](https://circleci.com/gh/kazupon/poeditor-service-provider)

POEditor service provider for [vue-i18n-locale-message](https://github.com/kazupon/vue-i18n-locale-message)


## :cd: Installation

```bash
$ npm install -g poeditor-service-provider # or npm install --save-dev poeditor-service-provider
```

due to use this provider, you need to install vue-i18n-locale-message beforehand.

```bash
$ npm install -g vue-i18n-locale-message
```


## :rocket: Usages

### Configurations

Before you use this provider, you need to configure the following:

```json5
{
  "provider": {
    "id": "12345", // your POEditor project id
    "token": "xxx..." // your POEditor API token
  }
}
```

About details, See the [`POEditorProviderConfiguration`](https://github.com/kazupon/poeditor-service-provider/blob/master/types/index.d.ts).

### Push the locale messages to POEditor

```bash
$ vue-i18n-locale-message push --provider=poeditor-service-provider \
    --conf ./poeditor-service-provider-conf.json \
    --target-paths=./src/locales/*.json \
    --filename-match=^([\\w]*)\\.json
```

### Pull the locale messages from POEditor

```bash
$ vue-i18n-locale-message pull --provider=poeditor-service-provider \
    --conf ./poeditor-service-provider-conf.json \
    --output=./src/locales
```

### Indicate translation status from POEditor

```bash
$ vue-i18n-locale-message status --provider=poeditor-service-provider \
  --conf=poeditor-service-provider-conf.json
```

### Diff locale messages between local and POEditor

```bash
vue-i18n-locale-message diff --provider=poeditor-service-provider \
  --conf=poeditor-service-provider-conf.json \
  --target-paths=./src/locales/*.json \
  --filename-match=^([\\w]*)\\.json
```

### Import the locale messages to POEditor

```bash
$ vue-i18n-locale-message import --provider=poeditor-service-provider \
    --conf ./poeditor-service-provider-conf.json \
    --target=./src/locales/ja.json \
    --format=json
```

### Export the locale messages from POEditor

```bash
$ vue-i18n-locale-message export --provider=poeditor-service-provider \
    --conf ./poeditor-service-provider-conf.json \
    --output=./src/locales
```


## :warning: Do you have a hierarchical locale message?

POEditor will process locale messages with hierarchical structure as `context`.

Therefore, we need to normalize with flat structure , and push it to POEditor.

```bash
$ vue-i18n-locale-message push --provider=poeditor-service-provider \
    --conf ./poeditor-service-provider-conf.json \
    --target-paths=./src/locales/*.json \
    --nomalize=flat \
    --filename-match=^([\\w]*)\\.json
```

And also, when pulling data from POEditor, it need to normalize from flat structure to hierarchical structure.

```bash
$ vue-i18n-locale-message pull --provider=poeditor-service-provider \
    --conf ./poeditor-service-provider-conf.json \
    --nomalize=hierarchy \
    --output=./src/locales
```

## :copyright: License

[MIT](http://opensource.org/licenses/MIT)


CHANGELOG.md

## v0.7.1 (2020-02-05)

#### :bug: Bug Fixes
* [#53](https://github.com/kazupon/poeditor-service-provider/pull/53) fix object is converted to array ([@kazupon](https://github.com/kazupon))

#### Committers: 1
- kazuya kawaguchi ([@kazupon](https://github.com/kazupon))


## v0.7.0 (2020-01-09)

#### :star: Features
* [#41](https://github.com/kazupon/poeditor-service-provider/pull/41) export method supporting ([@kazupon](https://github.com/kazupon))
* [#40](https://github.com/kazupon/poeditor-service-provider/pull/40) import method feature ([@kazupon](https://github.com/kazupon))

#### :zap: Improvement Features
* [#38](https://github.com/kazupon/poeditor-service-provider/pull/38) improve pull format option ([@kazupon](https://github.com/kazupon))

#### Committers: 1
- kazuya kawaguchi ([@kazupon](https://github.com/kazupon))


## v0.6.0 (2020-01-07)

#### :star: Features
* [#35](https://github.com/kazupon/poeditor-service-provider/pull/35) status feature ([@kazupon](https://github.com/kazupon))

#### Committers: 1
- kazuya kawaguchi ([@kazupon](https://github.com/kazupon))


## v0.5.0 (2019-12-27)

#### :boom: Breaking Change
* [#30](https://github.com/kazupon/poeditor-service-provider/pull/30) breaking: fix push mode ([@kazupon](https://github.com/kazupon))

#### Committers: 1
- kazuya kawaguchi ([@kazupon](https://github.com/kazupon))


## v0.4.0 (2019-12-26)

#### :star: Features
* [#27](https://github.com/kazupon/poeditor-service-provider/pull/27) locale messages normalization feature ([@kazupon](https://github.com/kazupon))

#### Committers: 1
- kazuya kawaguchi ([@kazupon](https://github.com/kazupon))


## v0.3.0 (2019-12-24)

#### :star: Features
* [#23](https://github.com/kazupon/poeditor-service-provider/pull/23) feat: support for API token from ENV ([@kazupon](https://github.com/kazupon))

#### :zap: Improvement Features
* [#24](https://github.com/kazupon/poeditor-service-provider/pull/24) Token omittion on the configuration ([@kazupon](https://github.com/kazupon))
* [#22](https://github.com/kazupon/poeditor-service-provider/pull/22) update provider interface ([@kazupon](https://github.com/kazupon))

#### Committers: 1
- kazuya kawaguchi ([@kazupon](https://github.com/kazupon))


## v0.2.0 (2019-12-20)

#### :star: Features
* [#18](https://github.com/kazupon/poeditor-service-provider/pull/18) pull locale message feature ([@kazupon](https://github.com/kazupon))

#### Committers: 1
- kazuya kawaguchi ([@kazupon](https://github.com/kazupon))


## v0.1.1 (2019-12-18)

#### :zap: Improvement Features
* [#13](https://github.com/kazupon/poeditor-service-provider/pull/13) fix unneccesary log ([@kazupon](https://github.com/kazupon))

#### Committers: 1
- kazuya kawaguchi ([@kazupon](https://github.com/kazupon))


## v0.1.0 (2019-12-18)

#### :star: Features
* [#10](https://github.com/kazupon/poeditor-service-provider/pull/10) push resource feature ([@kazupon](https://github.com/kazupon))

#### :pencil: Documentation
* [#11](https://github.com/kazupon/poeditor-service-provider/pull/11) add usage section ([@kazupon](https://github.com/kazupon))

#### Committers: 1
- kazuya kawaguchi ([@kazupon](https://github.com/kazupon))

## v0.0.0 (2019-12-18)

I've borned in the world! :tada:

#### Committers: 1
- kazuya kawaguchi ([@kazupon](https://github.com/kazupon))


test/api.test.ts
import * as path from 'path'

// ------
// mocks
jest.mock('axios')
import axios from 'axios'
import {
  getLocales,
  getTranslationStatus,
  exportLocaleMessage,
  exportRawLocaleMessage,
  upload
} from '../src/api'

// --------------------
// setup/teadown hooks

let spyLog
beforeEach(() => {
  spyLog = jest.spyOn(global.console, 'log')
})

afterEach(() => {
  spyLog.mockRestore()
  jest.clearAllMocks()
})

// -----------
// test cases

test('getLocales', async () => {
  // mocking ...
  const axiosMock = axios as jest.Mocked<typeof axios>
  axiosMock.post.mockImplementationOnce((url, data, config) => Promise.resolve({
    data: {result: { languages: [{
      code: 'en',
      percentage: 1000.00
    }, {
      code: 'ja',
      percentage: 724.00
    }]}}
  }))
  
  // run
  const locales = await getLocales({ token: 'xxx', id: '12345' })

  // verify
  expect(locales).toEqual(['en', 'ja'])
})

test('getTranslationStatus', async () => {
  // mocking ...
  const axiosMock = axios as jest.Mocked<typeof axios>
  axiosMock.post.mockImplementationOnce((url, data, config) => Promise.resolve({
    data: {result: { languages: [{
      code: 'en',
      percentage: 1000.00
    }, {
      code: 'ja',
      percentage: 724.00
    }]}}
  }))
  
  // run
  const status = await getTranslationStatus({ token: 'xxx', id: '12345' })

  // verify
  expect(status).toEqual([{
    locale: 'en',
    percentage: 1000.00
  }, {
    locale: 'ja',
    percentage: 724.00
  }])
})

test('exportRawLocaleMessage', async () => {
  // mocking ...
  const DONWLOAD_URL = 'https://intlify.dev/data'
  const data = {
    term: 'hello',
    definition: 'hello'
  }
  const axiosMock = axios as jest.Mocked<typeof axios>
  axiosMock.post.mockImplementationOnce((url, data, config) => Promise.resolve({
    data: { result: { url: DONWLOAD_URL } }
  }))
  axiosMock.get.mockImplementationOnce(url => Promise.resolve({ data: JSON.stringify(data) }))

  const messages = await exportRawLocaleMessage({ token: 'xxx', id: '12345' }, 'en', 'json')

  expect(messages).toEqual({
    locale: 'en',
    format: 'json',
    data: Buffer.from(JSON.stringify(data))
  })
  expect(axiosMock.get).toHaveBeenCalledTimes(1)
  expect(axiosMock.get).toHaveBeenLastCalledWith(DONWLOAD_URL)
})

test('exportLocaleMessage', async () => {
  // mocking ...
  const DONWLOAD_URL = 'https://intlify.dev/locales'
  const axiosMock = axios as jest.Mocked<typeof axios>
  axiosMock.post.mockImplementationOnce((url, data, config) => Promise.resolve({
    data: { result: { url: DONWLOAD_URL } }
  }))
  axiosMock.get.mockImplementationOnce(url => Promise.resolve({
    data: [{
      term: 'hello',
      definition: 'hello'
    }]
  }))

  const messages = await exportLocaleMessage({ token: 'xxx', id: '12345' }, 'en', 'json')

  expect(messages).toEqual([{
    definition: 'hello',
    term: 'hello'
  }])
  expect(axiosMock.get).toHaveBeenCalledTimes(1)
  expect(axiosMock.get).toHaveBeenLastCalledWith(DONWLOAD_URL)
})

test('upload', async () => {
  // mocking ...
  const axiosMock = axios as jest.Mocked<typeof axios>
  axiosMock.post.mockImplementation((url, data, config) => Promise.resolve({ data: { result: {} } }))

  // run
  const res = await upload({
    locale: 'en',
    path: path.resolve('./test/fixtures/en.json')
  }, {
    token: 'xxx',
    id: '12345'
  })

  // verify
  expect(res).toEqual({ result: {} })
})


test/pull.test.ts
import factory from '../src/index'
import { ProviderConfiguration } from 'vue-i18n-locale-message'
import { POEditorProviderConfiguration } from '../types'

// ------
// mocks
jest.mock('axios')
import axios from 'axios'

// --------------------
// setup/teadown hooks

let spyLog
beforeEach(() => {
  spyLog = jest.spyOn(global.console, 'log')
})

afterEach(() => {
  spyLog.mockRestore()
  jest.clearAllMocks()
})

// -----------
// test cases

test('specify locales', async () => {
  // mocking ...
  const DONWLOAD_URL = 'https://intlify.dev/locales'
  const axiosMock = axios as jest.Mocked<typeof axios>
  axiosMock.post.mockImplementationOnce((url, data, config) => Promise.resolve({
    data: { result: { url: DONWLOAD_URL } }
  }))
  axiosMock.get.mockImplementationOnce(url => Promise.resolve({
    data: [{
      term: 'hello',
      definition: 'hello'
    }]
  }))
  axiosMock.post.mockImplementationOnce((url, data, config) => Promise.resolve({
    data: { result: { url: DONWLOAD_URL } }
  }))
  axiosMock.get.mockImplementationOnce(url => Promise.resolve({
    data: [{
      term: 'hello',
      definition: 'こんにちわわわ！'
    }]
  }))
  
  // run
  const conf: ProviderConfiguration<POEditorProviderConfiguration> = {
    provider: {
      token: 'xxx',
      id: '12345',
      interval: 1
    },
    pushMode: 'file-path'
  }
  const provider = factory(conf)
  const resource = await provider.pull({ locales: ['en', 'ja'], dryRun: false })

  // verify
  expect(spyLog).toHaveBeenNthCalledWith(1, `fetch 'en' locale messages`)
  expect(spyLog).toHaveBeenNthCalledWith(2, `fetch 'ja' locale messages`)
  expect(resource).toMatchObject({
    en: { hello: 'hello' },
    ja: { hello: 'こんにちわわわ！' }
  })
  expect(axiosMock.get).toHaveBeenCalledTimes(2)
  expect(axiosMock.get).toHaveBeenLastCalledWith(DONWLOAD_URL)
})

test('not specify locales', async () => {
  // mocking ...
  const DONWLOAD_URL = 'https://intlify.dev/locales'
  const axiosMock = axios as jest.Mocked<typeof axios>
  axiosMock.post.mockImplementationOnce((url, data, config) => Promise.resolve({
    data: {result: { languages: [{
      code: 'en'
    }, {
      code: 'ja'
    }]}}
  }))
  axiosMock.post.mockImplementationOnce((url, data, config) => Promise.resolve({
    data: { result: { url: DONWLOAD_URL } }
  }))
  axiosMock.get.mockImplementationOnce(url => Promise.resolve({
    data: [{
      term: 'hello',
      definition: 'hello'
    }]
  }))
  axiosMock.post.mockImplementationOnce((url, data, config) => Promise.resolve({
    data: { result: { url: DONWLOAD_URL } }
  }))
  axiosMock.get.mockImplementationOnce(url => Promise.resolve({
    data: [{
      term: 'hello',
      definition: 'こんにちわわわ！'
    }]
  }))
  
  // run
  const conf: ProviderConfiguration<POEditorProviderConfiguration> = {
    provider: {
      token: 'xxx',
      id: '12345',
      interval: 1
    },
    pushMode: 'file-path'
  }
  const provider = factory(conf)
  const resource = await provider.pull({ locales: ['en', 'ja'], dryRun: true })

  // verify
  expect(spyLog).toHaveBeenNthCalledWith(1, '----- POEditorServiceProvider pull dryRun mode -----')
  expect(spyLog).toHaveBeenNthCalledWith(2, `fetch 'en' locale messages`)
  expect(spyLog).toHaveBeenNthCalledWith(3, `fetch 'ja' locale messages`)
  expect(resource).toMatchObject({
    en: { hello: 'hello' },
    ja: { hello: 'こんにちわわわ！' }
  })
  expect(axiosMock.get).toHaveBeenCalledTimes(2)
  expect(axiosMock.get).toHaveBeenLastCalledWith(DONWLOAD_URL)
})


test/push.test.ts
import * as path from 'path'
import factory from '../src/index'
import { ProviderConfiguration } from 'vue-i18n-locale-message'
import { POEditorProviderConfiguration } from '../types'

// ------
// mocks
jest.mock('axios')
import axios from 'axios'

// --------------------
// setup/teadown hooks

let spyLog
beforeEach(() => {
  spyLog = jest.spyOn(global.console, 'log')
})

afterEach(() => {
  spyLog.mockRestore()
  jest.clearAllMocks()
})

// -----------
// test cases

test('API call', async () => {
  // resources
  const enResource = path.resolve('./test/fixtures/en.json')
  const jaResource = path.resolve('./test/fixtures/ja.json')

  // mocking ...
  const axiosMock = axios as jest.Mocked<typeof axios>
  axiosMock.post.mockImplementation((url, data, config) => Promise.resolve({ data: {} }))

  // run
  const conf: ProviderConfiguration<POEditorProviderConfiguration> = {
    provider: {
      token: 'xxx',
      id: '12345',
      interval: 1
    },
    pushMode: 'file-path'
  }
  const provider = factory(conf)
  await provider.push({
    resource: {
      mode: conf.pushMode,
      files: [{
        locale: 'en',
        path: path.resolve('./test/fixtures/en.json')
      }, {
        locale: 'ja',
        path: path.resolve('./test/fixtures/ja.json')
      }]
    },
    dryRun: false
  })

  // verify
  expect(spyLog).toHaveBeenNthCalledWith(1, `upload '${enResource}' file with 'en' locale`)
  expect(spyLog).toHaveBeenNthCalledWith(2, 'wait 30 sec due to limit Editor API call ...')
  expect(spyLog).toHaveBeenNthCalledWith(3, `upload '${jaResource}' file with 'ja' locale`)
  expect(spyLog).toHaveBeenNthCalledWith(4, 'wait 30 sec due to limit Editor API call ...')
})

test('dryRun mode', async () => {
  // run
  const conf: ProviderConfiguration<POEditorProviderConfiguration> = {
    provider: {
      token: 'xxx',
      id: '12345'
    },
    pushMode: 'file-path'
  }
  const provider = factory(conf)
  await provider.push({
    resource: {
      mode: conf.pushMode,
      files: [{
        locale: 'en',
        path: '/path/to/project/src/locales/en.json'
      }]
    },
    dryRun: true
  })

  // verify
  expect(spyLog).toHaveBeenNthCalledWith(1, '----- POEditorServiceProvider push dryRun mode -----')
  expect(spyLog).toHaveBeenNthCalledWith(2, "upload '/path/to/project/src/locales/en.json' file with 'en' locale")
})


src/api.ts
import * as fs from 'fs'
import qs, { ParsedUrlQueryInput } from 'querystring'
import FormData from 'form-data'
import axios from 'axios'

import { debug as Debug } from 'debug'
const debug = Debug('poeditor-service-provider:api')

import { Locale, TranslationStatus, RawLocaleMessage } from 'vue-i18n-locale-message'
import {
  POEditorProviderConfiguration,
  POEditorLocaleMessage,
  POEditorLanguage,
  UploadFileInfo
} from '../types'

const POEDITOR_API_BASE_URL = 'https://api.poeditor.com/v2'

async function getLanguageList (config: POEditorProviderConfiguration) {
  return new Promise<POEditorLanguage[]>((resolve, reject) => {
    if (!config.token) { return reject(new Error('invalid `config.token` param')) }
    const data: ParsedUrlQueryInput = {
      api_token: config.token, // eslint-disable-line
      id: config.id
    }
    const headers = { 'Content-Type': 'application/x-www-form-urlencoded' }
    axios.post(`${POEDITOR_API_BASE_URL}/languages/list`, qs.stringify(data), { headers })
      .then(res => {
        const languages = res.data.result.languages as POEditorLanguage[]
        debug('fetch languages:', languages)
        resolve(languages)
      })
  })
}

export async function getLocales (config: POEditorProviderConfiguration) {
  const languages = await getLanguageList(config)
  return languages.map(lang => lang.code as Locale) as Locale[]
}

export async function getTranslationStatus (config: POEditorProviderConfiguration) {
  const languages = await getLanguageList(config)
  return languages.map(lang => ({ locale: lang.code, percentage: lang.percentage })) as TranslationStatus[]
}

export async function exportRawLocaleMessage (config: POEditorProviderConfiguration, locale: Locale, format: string) {
  const download = async (url: string): Promise<any> => { // eslint-disable-line
    return new Promise(resolve => {
      axios.get(url)
        .then(res => {
          debug('donwload:', res.data)
          resolve(res.data)
        })
    })
  }

  return new Promise<RawLocaleMessage>((resolve, reject) => {
    if (!config.token) { return reject(new Error('invalid `config.token` param')) }
    const data: ParsedUrlQueryInput = {
      api_token: config.token, // eslint-disable-line
      id: config.id,
      language: locale,
      type: format
    }
    const headers = { 'Content-Type': 'application/x-www-form-urlencoded' }
    axios.post(`${POEDITOR_API_BASE_URL}/projects/export`, qs.stringify(data), { headers })
      .then(async (res) => {
        const url = res.data.result.url as string
        debug('/projects/export url:', url)
        const raw = await download(url)
        resolve({
          locale,
          format,
          data: Buffer.from(raw, 'utf-8')
        })
      })
  })
}

export async function exportLocaleMessage (config: POEditorProviderConfiguration, locale: Locale, format: string) {
  const download = async (url: string): Promise<POEditorLocaleMessage[]> => {
    return new Promise(resolve => {
      axios.get(url)
        .then(res => {
          debug('donwload:', res.data)
          resolve(res.data as POEditorLocaleMessage[])
        })
    })
  }

  return new Promise<POEditorLocaleMessage[]>((resolve, reject) => {
    if (!config.token) { return reject(new Error('invalid `config.token` param')) }
    const data: ParsedUrlQueryInput = {
      api_token: config.token, // eslint-disable-line
      id: config.id,
      language: locale,
      type: format
    }
    const headers = { 'Content-Type': 'application/x-www-form-urlencoded' }
    axios.post(`${POEDITOR_API_BASE_URL}/projects/export`, qs.stringify(data), { headers })
      .then(async (res) => {
        const url = res.data.result.url as string
        debug('/projects/export url:', url)
        const raw = await download(url)
        resolve(raw)
      })
  })
}

export async function upload (fileInfo: UploadFileInfo, config: POEditorProviderConfiguration) {
  return new Promise(resolve => {
    const file = fs.createReadStream(fileInfo.path)
    const data = new FormData()
    data.append('file', file)
    data.append('api_token', config.token)
    data.append('id', config.id)
    data.append('language', fileInfo.locale)
    data.append('updating', 'terms_translations')
    data.append('overwrite', '1')
    data.append('sync_terms', '1')
    const formHeader = data.getHeaders()
    axios
      .post(`${POEDITOR_API_BASE_URL}/projects/upload`, data, { headers: { ...formHeader }})
      .then(ret => resolve(ret.data))
  })
}


src/utils.ts
import * as fs from 'fs'
import * as path from 'path'
import { promisify } from 'util'
import { flatten } from 'flat'
import tmp from 'tmp-promise'

import { UploadFileInfo } from '../types'
import { LocaleMessages, Locale, RawLocaleMessage } from 'vue-i18n-locale-message'

const readFilePromisify = promisify(fs.readFile)
const writeFilePromisify = promisify(fs.writeFile)

export function getToken (defaultToken?: string) {
  return process.env.POEDITOR_API_TOKEN || defaultToken
}

export const delay = (sec: number) => {
  return new Promise(resolve => { setTimeout(resolve, sec * 1000) })
}

async function saveToTmp (messages: LocaleMessages, indent: number, dryRun: boolean) {
  const files = [] as UploadFileInfo[]
  let dir = null
  try {
    dir = await tmp.dir()
    const locales = Object.keys(messages) as Locale[]
    for (const locale of locales) {
      const tmpFilePath = path.resolve(dir.path, `${locale}.json`)
      console.log(`save locale message to tmp: ${tmpFilePath}`)
      if (!dryRun) {
        await writeFilePromisify(tmpFilePath, JSON.stringify(messages[locale], null, indent))
      }
      files.push({ locale: locale, path: tmpFilePath })
    }
    return Promise.resolve(files)
  } finally {
    dryRun && dir?.cleanup()
  }
}

export async function getUploadFiles (messages: LocaleMessages, indent: number, dryRun: boolean, normalize?: string) {
  let dir = null
  try {
    const files = await saveToTmp(messages, indent, dryRun)
    if (!normalize) {
      return Promise.resolve(files)
    } else {
      dir = await tmp.dir()
      const normalizedFiles = [] as UploadFileInfo[]
      for (const file of files) {
        const orgData = await readFilePromisify(file.path)
        const orgJSON = JSON.parse(orgData.toString())
        const frattedJSON = flatten(orgJSON)
        const parsedOrgFile = path.parse(file.path)
        const tmpFilePath = path.resolve(dir.path, parsedOrgFile.base)
        console.log(`normalize locale messages: ${file.path} -> ${tmpFilePath}`)
        if (!dryRun) {
          await writeFilePromisify(tmpFilePath, JSON.stringify(frattedJSON, null, indent))
        }
        normalizedFiles.push({ locale: file.locale, path: tmpFilePath })
      }
      return Promise.resolve(normalizedFiles)
    }
  } finally {
    dryRun && dir?.cleanup()
  }
}

export async function getUploadFilesWithRaw (messages: RawLocaleMessage[], dryRun: boolean) {
  const files = [] as UploadFileInfo[]
  let dir = null
  try {
    dir = await tmp.dir()
    for (const { locale, format, data } of messages) {
      const tmpFilePath = path.resolve(dir.path, `${locale}.${format}`)
      console.log(`save raw locale message to tmp: ${tmpFilePath}`)
      if (!dryRun) {
        await writeFilePromisify(tmpFilePath, data)
      }
      files.push({ locale, path: tmpFilePath })
    }
    return Promise.resolve(files)
  } finally {
    dryRun && dir?.cleanup()
  }
}


src/provider.ts
import { unflatten } from 'flat'
import { delay, getUploadFiles, getUploadFilesWithRaw } from './utils'
import {
  Locale,
  Provider,
  PushArguments,
  PullArguments,
  StatusArguments,
  ImportArguments,
  ExportArguments,
  RawLocaleMessage,
  LocaleMessage,
  LocaleMessages,
  TranslationStatus
} from 'vue-i18n-locale-message'
import {
  upload,
  getLocales,
  getTranslationStatus,
  exportRawLocaleMessage,
  exportLocaleMessage
} from './api'

import { debug as Debug } from 'debug'
const debug = Debug('poeditor-service-provider:provider')

const POEDITOR_API_INTERVAL_LIMITATION = 30

export default function provider (
  id: string,
  token: string,
  interval: number,
  indent: number
): Provider {
  /**
   *  push
   */
  const push = async (args: PushArguments): Promise<void> => {
    const { messages, dryRun, normalize } = args

    if (normalize && normalize !== 'flat') {
      return Promise.reject(new Error(`support nomalization format 'flat' only`))
    }

    dryRun && console.log(`----- POEditorServiceProvider push dryRun mode -----`)
    const results = []
    const files = await getUploadFiles(messages, indent, dryRun, normalize)

    for (const file of files) {
      console.log(`upload '${file.locale}' locale`)
      if (!dryRun) {
        const ret = await upload(file, { token, id })
        debug(`upload file '${file.path}' result`, ret)
        console.log(`wait ${POEDITOR_API_INTERVAL_LIMITATION} sec due to limit Editor API call ...`)
        await delay(interval)
        results.push(ret)
      } else {
        await delay(1)
        results.push(undefined)
      }
    }

    return Promise.resolve()
  }

  /**
   *  pull
   */
  const pull = async (args: PullArguments): Promise<LocaleMessages> => {
    const { locales, dryRun, normalize, format } = args

    dryRun && console.log(`----- POEditorServiceProvider pull dryRun mode -----`)
    const messages = {} as LocaleMessages

    if (normalize && normalize !== 'hierarchy') {
      return Promise.reject(new Error(`support nomalization format 'hierarchy' only`))
    }

    const fetchLocales = async (locales: Locale[]) => {
      if (locales.length === 0) {
        console.log('fetch locales')
        return await getLocales({ token, id })
      } else {
        return Promise.resolve(locales)
      }
    }

    const targetLocales = await fetchLocales(locales)
    for (const locale of targetLocales) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      let message = {} as any // TODO: should be refactored LocaleMessage type definition
      console.log(`fetch '${locale}' locale messages`)
      const poeditorLocaleMessages = await exportLocaleMessage({ token, id }, locale, format)
      message = poeditorLocaleMessages.reduce((mesasge, m) => {
        message[m.term] = m.definition || ''
        return mesasge
      }, message)
      messages[locale] = (!normalize ? message : unflatten(message, { object: true })) as LocaleMessage
    }
    debug('fetch locale messages', messages)

    return Promise.resolve(messages)
  }

  /**
   *  status
   */
  const status = async (args: StatusArguments): Promise<TranslationStatus[]> => {
    const { locales } = args
    const translationStatus = await getTranslationStatus({ token, id })
    return Promise.resolve(
      locales.length === 0
        ? translationStatus
        : translationStatus.filter(st => locales.includes(st.locale))
    )
  }

  /**
   *  import
   */
  const _import = async (args: ImportArguments): Promise<void> => {
    const { messages, dryRun } = args

    dryRun && console.log(`----- POEditorServiceProvider import dryRun mode -----`)
    const results = []
    const files = await getUploadFilesWithRaw(messages, dryRun)

    for (const file of files) {
      console.log(`upload '${file.locale}' locale`)
      if (!dryRun) {
        const ret = await upload(file, { token, id })
        debug(`upload file '${file.path}' result`, ret)
        console.log(`wait ${POEDITOR_API_INTERVAL_LIMITATION} sec due to limit Editor API call ...`)
        await delay(interval)
        results.push(ret)
      } else {
        await delay(1)
        results.push(undefined)
      }
    }

    return Promise.resolve()
  }

  /**
   *  export
   */
  const _export = async (args: ExportArguments): Promise<RawLocaleMessage[]> => {
    const { locales, dryRun, format } = args

    dryRun && console.log(`----- POEditorServiceProvider export dryRun mode -----`)
    const messages = [] as RawLocaleMessage[]

    const fetchLocales = async (locales: Locale[]) => {
      if (locales.length === 0) {
        console.log('fetch locales')
        return await getLocales({ token, id })
      } else {
        return Promise.resolve(locales)
      }
    }

    const targetLocales = await fetchLocales(locales)
    for (const locale of targetLocales) {
      console.log(`fetch '${locale}' raw locale messages`)
      const message = await exportRawLocaleMessage({ token, id }, locale, format)
      messages.push(message)
    }
    debug('raw locale messages', messages)

    return Promise.resolve(messages)
  }

  return { push, pull, status, import: _import, export: _export } as Provider
}


types/index.d.ts
import { Locale } from 'vue-i18n-locale-message'

/**
 *  POEditor Provider configuration for vue-i18n-locale-message CLI --conf option
 *  e.g.
 *    {
 *      "provider": {
 *        "token": "xxx",
 *        "id": "12345"
 *      }
 *    }
 */
export type POEditorProviderConfiguration = {
  /**
   *  project id
   */
  id: string
  /**
   *  API token.
   *  if it's ommitted, use the value of `POEDITOR_API_TOKEN` ENV.
   */
  token?: string
  /**
   *  API call interval (due to limit for API calling, see https://poeditor.com/docs/api_rates).
   *  if it's omitted, internally set 30 sec as default.
   */
  interval?: number
  /**
   *  Normalize locale messages file indend, default indent 2 space
   */
  indent?: number
}

export type POEditorLocaleMessage = {
  term: string
  definition: string
  context: string
  term_plural: string
  reference: string
  comment: string
}

export type POEditorLanguage = {
  name: string
  code: string
  translations: number
  percentage: number
}

export type UploadFileInfo = {
  locale: Locale
  path: string
}


src/index.ts
import * as fs from 'fs'
import qs from 'querystring'
import FormData from 'form-data'
import axios from 'axios'
import { getToken } from './utils'

import {
  Locale,
  Provider,
  PushArguments,
  PullArguments,
  ProviderPullResource,
  ProviderPushFileInfo,
  ProviderConfiguration,
  LocaleMessage
} from 'vue-i18n-locale-message'
import { POEditorProviderConfiguration, POEditorLocaleMessage } from '../types'

import { debug as Debug } from 'debug'
const debug = Debug('poeditor-service-provider')

const POEDITOR_API_BASE_URL = 'https://api.poeditor.com/v2'
const POEDITOR_API_INTERVAL_LIMITATION = 30

async function getLocales (config: POEditorProviderConfiguration) {
  return new Promise<Locale[]>(resolve => {
    const data = {
      api_token: config.token,
      id: config.id
    }
    const headers = { 'Content-Type': 'application/x-www-form-urlencoded' }
    axios.post(`${POEDITOR_API_BASE_URL}/languages/list`, qs.stringify(data), { headers })
      .then(res => {
        const locales = res.data.result.languages.map((lang: any) => lang.code as Locale) as Locale[]
        debug('fetch locales:', locales)
        resolve(locales)
      })
  })
}

async function exportLocaleMessage (config: POEditorProviderConfiguration, locale: Locale, format: string) {
  const download = async (url: string): Promise<POEditorLocaleMessage[]> => {
    return new Promise(resolve => {
      axios.get(url)
        .then(res => {
          debug('donwload:', res.data)
          resolve(res.data as POEditorLocaleMessage[])
        })
    })
  }

  return new Promise<POEditorLocaleMessage[]>(resolve => {
    const data = {
      api_token: config.token,
      id: config.id,
      language: locale,
      type: format
    }
    const headers = { 'Content-Type': 'application/x-www-form-urlencoded' }
    axios.post(`${POEDITOR_API_BASE_URL}/projects/export`, qs.stringify(data), { headers })
      .then(async (res) => {
        const url = res.data.result.url as string
        debug('/projects/export url:', url)
        const raw = await download(url)
        resolve(raw)
      })
  })
}

async function upload (fileInfo: ProviderPushFileInfo, config: POEditorProviderConfiguration) {
  return new Promise(resolve => {
    const file = fs.createReadStream(fileInfo.path)
    const data = new FormData()
    data.append('file', file)
    data.append('api_token', config.token)
    data.append('id', config.id)
    data.append('language', fileInfo.locale)
    data.append('updating', 'terms_translations')
    data.append('overwrite', '1')
    data.append('sync_terms', '1')
    const formHeader = data.getHeaders()
    axios
      .post(`${POEDITOR_API_BASE_URL}/projects/upload`, data, { headers: { ...formHeader }})
      .then(ret => resolve(ret.data))
  })
}

const factory = (configration: ProviderConfiguration<POEditorProviderConfiguration>): Provider => {
  const id = configration.provider.id
  const token = getToken(configration.provider.token)
  const interval = configration.provider.interval || 30
  const { pushMode } = configration
  const delay = (sec: number) => new Promise(resolve => { setTimeout(resolve, sec * 1000) })

  const push = async (args: PushArguments): Promise<void> => {
    const { resource, dryRun } = args
    return new Promise(async (resolve, reject) => {
      try {
        debug('provider#push:', resource, dryRun)
        debug(`push mode: config.pushMode=${pushMode}, resouroce.mode=${resource.mode}`)
        if (resource.mode !== pushMode && pushMode !== 'file-path') {
          throw new Error('invalid push mode!')
        }
        dryRun && console.log(`----- POEditorServiceProvider push dryRun mode -----`)
        const results = []
        const files = resource.files || []
        for (const file of files) {
          console.log(`upload '${file.path}' file with '${file.locale}' locale`)
          if (!dryRun) {
            const ret = await upload(file, { token, id })
            debug(`upload file '${file.path}' result`, ret)
            console.log(`wait ${POEDITOR_API_INTERVAL_LIMITATION} sec due to limit Editor API call ...`)
            await delay(interval)
            results.push(ret)
          } else {
            await delay(1)
            results.push(undefined)
          }
        }
        resolve()
      } catch (e) {
        reject(e)
      }
    })
  }

  const pull = async (args: PullArguments): Promise<ProviderPullResource> => {
    const { locales, dryRun } = args
    return new Promise(async (resolve, reject) => {
      dryRun && console.log(`----- POEditorServiceProvider pull dryRun mode -----`)
      const resource = {} as ProviderPullResource

      const fetchLocales = async (locales: Locale[]) => {
        if (locales.length === 0) {
          console.log('fetch locales')
          return await getLocales({ token, id })
        } else {
          return Promise.resolve(locales)
        }
      }

      try {
        const targetLocales = await fetchLocales(locales)
        for (const locale of targetLocales) {
          let message = {} as any // TODO: should be refactored LocaleMessage type definition
          console.log(`fetch '${locale}' locale messages`)
          const poeditorLocaleMessages = await exportLocaleMessage({ token, id }, locale, 'json')
          message = poeditorLocaleMessages.reduce((mesasge, m) => {
            message[m.term] = m.definition || ''
            return mesasge
          }, message)
          resource[locale] = message as LocaleMessage
        }
        debug('normalized resource', resource)

        resolve(resource)
      } catch (e) {
        reject(e)
      }
    })
  }

  return { push, pull } as Provider
}

export default factory
